Cookie not attached when request to server is redirected

### What is your Scenario?

When page is using, in my case oauth (or in other cases any redirection targeting tested service), request for authentication in identity manager is redirected from service asking for authentication.
User is logged in in identity manager and such redirected reques should respond with ok, when user has valid session.
Session cookie is httpOnly: true, secure: true.

Everything works as expected when test is executed on https domain (e.g. prod env).
However when I try to run it against http localhost on dev server it fails because cookie is not attached.

Page is written in angular 11 and dev server is started using angular's option to proxy traffic to the backend service,
with termination of ssl and translation of host name.

When using dev server manually, everything works fine as localhost is treated by the browser as safe location
and secure cookies are working correctly.

### What is the Current behavior?

```
fill data & click login ---------- /login ---------> Identity Manager
                        <---- OK, session cookie----

auth in backend service ---------- /auth  ---------> Backend service
using Identity Manager
                        <---- 302, /verify ---------

                        ---------- /verify --------> Identity Manager
                                without cookie!
                        <--------- 401 -------------
```

### What is the Expected behavior?

```
fill data & click login ---------- /login ---------> Identity Manager
                        <---- OK, session cookie----

auth in backend service ---------- /auth  ---------> Backend service
using Identity Manager
                        <---- 302, /verify ---------

                        - /verify, session cookie -> Identity Manager
                        <--------- 200 -------------
```

### What is your public website URL? (or attach your complete example)

```js
var express = require('express');
var cookieParser = require('cookie-parser')

var app = express();

app.use(cookieParser());

app.use(function (req, res, next) {
  next();
  console.log(`<<< ${req.method} ${req.originalUrl}`);
  console.log(`>>> ${res.statusCode}\n`);
});

app.listen(3000);

app.post('/login', (req, res) => {
  res.cookie('session', 'abcd', {
    maxAge: 86400 * 1000,
    httpOnly: true,
    secure: true
  });
  res.status(200);
  res.send();
});

app.post('/logout', (req, res) => {
  res.clearCookie("session");
  res.status(200);
  res.send();
});

app.get('/authorize', (req, res) => {
  res.redirect(302, '/check-auth');
});

app.get('/check-auth', (req, res) => {
  const s = req.cookies && req.cookies.session ? req.cookies.session : '';
  if (s === 'abcd') {
    res.status(200);
  } else {
    res.status(401);
  }
  res.send();
});

app.get('/', (req, res) => {
  res.send(`
  <html>
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
  <script>
    let domain = "http://localhost:3000";
    function login() {
      $.ajax({
          url: domain + "/login",
          method: "post",
          complete(res) { console.log("Login: ", res.status); showResult(res.status); }
      })
    }
    function logout() {
      $.ajax({
          url: domain + "/logout",
          method: "post",
          complete(res) { console.log("Logout: ", res.status); showResult(res.status); }
      })
    }
    function auth() {
      $.ajax({
          url: domain + "/authorize",
          complete(res) { console.log("Auth: ", res.status); showResult(res.status); }
      })
    }
    function showResult(res) {
      if (res === 200) {
        $('#ok').show();
        $('#fail').hide();
      } else {
        $('#ok').hide();
        $('#fail').show();
      }
    }
  </script>
  <body>
  <button id="login" onclick="login()">login</button>
  <button id="auth" onclick="auth()">auth</button>
  <button id="logout" onclick="logout()">logout</button>
  </br>
  <h1 id="ok" style="color:#0f0; display: none">OK</h1>
  <h1 id="fail" style="color:#f00; display: none">FAIL</h1>
  </body>
  </html>
  `)
});

```

### What is your TestCafe test code?

```js
import { Selector } from 'testcafe';

fixture('Login Page')
  .page('http://localhost:3000');

test('can login with valid account', async t => {

  await t
    .click('button#login')

    .expect(Selector('#ok').visible).ok()

    .click('button#auth')

    .expect(Selector('#ok').visible).ok();
});
```

### Your complete configuration file

default configuration without any changes

### Your complete test report

 Running tests in:
 - Firefox 94.0 / Linux 0.0

 Login Page
 ✖ can login with valid account

   1) AssertionError: expected false to be truthy

      Browser: Firefox 94.0 / Linux 0.0

         10 |
         11 |    .expect(Selector('#ok').visible).ok()
         12 |
         13 |    .click('button#auth')
         14 |
       \> 15 |    .expect(Selector('#ok').visible).ok();
         16 |});
         17 |

         at <anonymous> (/home/ac/web/e2e/test.e2e.ts:15:38)
         at <anonymous> (/home/ac/web/e2e/test.e2e.ts:8:71)
         at __awaiter (/home/ac/web/e2e/test.e2e.ts:4:12)
         at <anonymous> (/home/ac/web/e2e/test.e2e.ts:6:48)



 1/1 failed (6s)


### Screenshots

N/A

### Steps to Reproduce

1.Run example server
2.Test manually - press login, press auth, both gave ok result
3.Run test, test fails


### TestCafe version

1.17.1

### Node.js version

v14.17.5

### Command-line arguments

npx testcafe firefox e2e/test.e2e.ts

### Browser name(s) and version(s)

Firefox 94.0, Chrome 95.0.4638.69

### Platform(s) and version(s)

Linux Manjaro 21.1.6

### Other

_No response_

--- a/src/request-pipeline/context.ts
+++ b/src/request-pipeline/context.ts
@@ -30,7 +30,7 @@ import getMockResponse from '../request-pipeline/request-hooks/response-mock/get
 import { Http2Response } from './destination-request/http2';
 import RequestEvent from '../session/events/request-event';
 
-interface DestInfo {
+export interface DestInfo {
     url: string;
     protocol: string;
     host: string;

--- a/src/request-pipeline/header-transforms/transforms.ts
+++ b/src/request-pipeline/header-transforms/transforms.ts
@@ -103,12 +103,12 @@ export const requestTransforms = {
     [BUILTIN_HEADERS.ifModifiedSince]:    skipIfStateSnapshotIsApplied,
     [BUILTIN_HEADERS.ifNoneMatch]:        skipIfStateSnapshotIsApplied,
     [BUILTIN_HEADERS.authorization]:      transformAuthorizationHeader,
-    [BUILTIN_HEADERS.proxyAuthorization]: transformAuthorizationHeader
+    [BUILTIN_HEADERS.proxyAuthorization]: transformAuthorizationHeader,
 };
 
 export const forcedRequestTransforms = {
     [BUILTIN_HEADERS.cookie]: (_src: string, ctx: RequestPipelineContext) =>
-        shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest) || void 0
+                                  shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest) || void 0,
 };
 
 // Response headers
@@ -125,19 +125,19 @@ export const responseTransforms = {
     [BUILTIN_HEADERS.proxyAuthenticate]: addAuthenticatePrefix,
 
     [BUILTIN_HEADERS.accessControlAllowOrigin]: (_src: string, ctx: RequestPipelineContext) =>
-        ctx.isSameOriginPolicyFailed ? void 0 : ctx.getProxyOrigin(!!ctx.dest.reqOrigin),
+                                                    ctx.isSameOriginPolicyFailed ? void 0 : ctx.getProxyOrigin(!!ctx.dest.reqOrigin),
 
     // NOTE: Change the transform type if we have an iframe with an image as src,
     // because it was transformed to HTML with the image tag.
     [BUILTIN_HEADERS.contentType]: (src: string, ctx: RequestPipelineContext) =>
-        ctx.contentInfo.isIframeWithImageSrc || ctx.contentInfo.isTextPage ? 'text/html' : src,
+                                       ctx.contentInfo.isIframeWithImageSrc || ctx.contentInfo.isTextPage ? 'text/html' : src,
 
     [BUILTIN_HEADERS.contentLength]: (src: string, ctx: RequestPipelineContext) =>
-        ctx.contentInfo.requireProcessing ? ctx.destResBody.length.toString() : src,
+                                         ctx.contentInfo.requireProcessing ? ctx.destResBody.length.toString() : src,
 
     // NOTE: We should skip an invalid trailer header (GH-2692).
     [BUILTIN_HEADERS.trailer]: (src: string, ctx: RequestPipelineContext) =>
-        ctx.destRes.headers[BUILTIN_HEADERS.transferEncoding] === 'chunked' ? src : void 0,
+                                   ctx.destRes.headers[BUILTIN_HEADERS.transferEncoding] === 'chunked' ? src : void 0,
 
     [BUILTIN_HEADERS.location]: (src: string, ctx: RequestPipelineContext) => {
         // NOTE: The RFC 1945 standard requires location URLs to be absolute. However, most popular browsers
@@ -160,7 +160,7 @@ export const responseTransforms = {
         src = src.replace('ALLOW-FROM', '').trim();
 
         const isCrossDomain = ctx.isIframe && !urlUtils.sameOriginCheck(ctx.dest.url, src);
-        const proxiedUrl = ctx.toProxyUrl(src, isCrossDomain, ctx.contentInfo.contentTypeUrlToken);
+        const proxiedUrl    = ctx.toProxyUrl(src, isCrossDomain, ctx.contentInfo.contentTypeUrlToken);
 
         return 'ALLOW-FROM ' + proxiedUrl;
     },
@@ -174,7 +174,7 @@ export const responseTransforms = {
             return void 0;
 
         return src;
-    }
+    },
 };
 
 export const forcedResponseTransforms = {

--- a/src/request-pipeline/header-transforms/transforms.ts
+++ b/src/request-pipeline/header-transforms/transforms.ts
@@ -103,12 +103,12 @@ export const requestTransforms = {
     [BUILTIN_HEADERS.ifModifiedSince]:    skipIfStateSnapshotIsApplied,
     [BUILTIN_HEADERS.ifNoneMatch]:        skipIfStateSnapshotIsApplied,
     [BUILTIN_HEADERS.authorization]:      transformAuthorizationHeader,
-    [BUILTIN_HEADERS.proxyAuthorization]: transformAuthorizationHeader,
+    [BUILTIN_HEADERS.proxyAuthorization]: transformAuthorizationHeader
 };
 
 export const forcedRequestTransforms = {
     [BUILTIN_HEADERS.cookie]: (_src: string, ctx: RequestPipelineContext) =>
-        shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest.url) || void 0,
+        shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest) || void 0
 };
 
 // Response headers
@@ -174,7 +174,7 @@ export const responseTransforms = {
             return void 0;
 
         return src;
-    },
+    }
 };
 
 export const forcedResponseTransforms = {

--- a/src/request-pipeline/header-transforms/transforms.ts
+++ b/src/request-pipeline/header-transforms/transforms.ts
@@ -108,7 +108,7 @@ export const requestTransforms = {
 
 export const forcedRequestTransforms = {
     [BUILTIN_HEADERS.cookie]: (_src: string, ctx: RequestPipelineContext) =>
-                                  shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest) || void 0,
+        shouldOmitCredentials(ctx) ? void 0 : ctx.session.cookies.getHeader(ctx.dest) || void 0,
 };
 
 // Response headers
@@ -125,19 +125,19 @@ export const responseTransforms = {
     [BUILTIN_HEADERS.proxyAuthenticate]: addAuthenticatePrefix,
 
     [BUILTIN_HEADERS.accessControlAllowOrigin]: (_src: string, ctx: RequestPipelineContext) =>
-                                                    ctx.isSameOriginPolicyFailed ? void 0 : ctx.getProxyOrigin(!!ctx.dest.reqOrigin),
+        ctx.isSameOriginPolicyFailed ? void 0 : ctx.getProxyOrigin(!!ctx.dest.reqOrigin),
 
     // NOTE: Change the transform type if we have an iframe with an image as src,
     // because it was transformed to HTML with the image tag.
     [BUILTIN_HEADERS.contentType]: (src: string, ctx: RequestPipelineContext) =>
-                                       ctx.contentInfo.isIframeWithImageSrc || ctx.contentInfo.isTextPage ? 'text/html' : src,
+        ctx.contentInfo.isIframeWithImageSrc || ctx.contentInfo.isTextPage ? 'text/html' : src,
 
     [BUILTIN_HEADERS.contentLength]: (src: string, ctx: RequestPipelineContext) =>
-                                         ctx.contentInfo.requireProcessing ? ctx.destResBody.length.toString() : src,
+        ctx.contentInfo.requireProcessing ? ctx.destResBody.length.toString() : src,
 
     // NOTE: We should skip an invalid trailer header (GH-2692).
     [BUILTIN_HEADERS.trailer]: (src: string, ctx: RequestPipelineContext) =>
-                                   ctx.destRes.headers[BUILTIN_HEADERS.transferEncoding] === 'chunked' ? src : void 0,
+        ctx.destRes.headers[BUILTIN_HEADERS.transferEncoding] === 'chunked' ? src : void 0,
 
     [BUILTIN_HEADERS.location]: (src: string, ctx: RequestPipelineContext) => {
         // NOTE: The RFC 1945 standard requires location URLs to be absolute. However, most popular browsers

--- a/src/session/cookies.ts
+++ b/src/session/cookies.ts
@@ -4,6 +4,7 @@ import { castArray, flattenDeep } from 'lodash';
 import { parseUrl } from '../utils/url';
 import { parse as parseJSON, stringify as stringifyJSON } from '../utils/json';
 import { URL } from 'url';
+import { DestInfo } from '../request-pipeline/context';
 
 const LOCALHOST_DOMAIN = 'localhost';
 const LOCALHOST_IP     = '127.0.0.1';
@@ -46,11 +47,11 @@ export default class Cookies {
         this._pendingSyncCookies   = [];
     }
 
-    _syncWrap(method: string) {
+    _syncWrap (method: string) {
         return (...args) => {
             let syncErr, syncResult;
             this._cookieJar.store[method](...args, (err, result) => {
-                syncErr = err;
+                syncErr    = err;
                 syncResult = result;
             });
 
@@ -82,8 +83,7 @@ export default class Cookies {
 
                 if (!cookie)
                     return resultCookies;
-            }
-            else
+            } else
                 cookie = cookieStr;
 
             // NOTE: If cookie.domain and url hostname are equal to localhost/127.0.0.1,
@@ -144,8 +144,8 @@ export default class Cookies {
     private _findCookiesByApi (urls: Url[], key?: string): (Cookie | Cookie[])[] {
         return urls.map(({ domain, path }) => {
             const cookies = key
-                            ? this._findCookieSync(domain, path, key)
-                            : this._findCookiesSync(domain, path);
+                ? this._findCookieSync(domain, path, key)
+                : this._findCookiesSync(domain, path);
 
             return cookies || [];
         });
@@ -244,7 +244,7 @@ export default class Cookies {
 
             for (const deletedCookie of deletedCookies) {
                 if (deletedCookie.domain && deletedCookie.path && deletedCookie.key) {
-                     this._removeCookieSync(deletedCookie.domain, deletedCookie.path, deletedCookie.key);
+                    this._removeCookieSync(deletedCookie.domain, deletedCookie.path, deletedCookie.key);
 
                     deletedCookie.expires = new Date(0);
                     this._pendingSyncCookies.push(deletedCookie);
@@ -278,7 +278,15 @@ export default class Cookies {
         return this._cookieJar.getCookieStringSync(url, { http: false });
     }
 
-    getHeader (url: string): string | null {
-        return this._cookieJar.getCookieStringSync(url, { http: true }) || null;
+    getHeader ({ url, hostname }: DestInfo): string | null {
+        // NOTE: https://github.com/DevExpress/testcafe-hammerhead/issues/2715
+        // Cookies with the secure attribute should be passed to the localhost server(without ssl) or any other server(with ssl).
+        // CookieJar only checks the protocol, but not a hostname. That is why we should to add secure attribute in case of localhost.
+        const cookieJarOpts =
+                  hostname === LOCALHOST_DOMAIN ||
+                  hostname === LOCALHOST_IP ?
+                      { http: true, secure: true } :
+                      { http: true };
+        return this._cookieJar.getCookieStringSync(url, cookieJarOpts) || null;
     }
 }

--- a/src/session/cookies.ts
+++ b/src/session/cookies.ts
@@ -83,7 +83,8 @@ export default class Cookies {
 
                 if (!cookie)
                     return resultCookies;
-            } else
+            }
+            else
                 cookie = cookieStr;
 
             // NOTE: If cookie.domain and url hostname are equal to localhost/127.0.0.1,
@@ -110,8 +111,8 @@ export default class Cookies {
 
     setJar (serializedJar): void {
         this._cookieJar = serializedJar
-            ? CookieJar.deserializeSync(parseJSON(serializedJar))
-            : new CookieJar();
+                          ? CookieJar.deserializeSync(parseJSON(serializedJar))
+                          : new CookieJar();
     }
 
     private _convertToExternalCookies (internalCookies: Cookie[]): ExternalCookies[] {
@@ -144,8 +145,8 @@ export default class Cookies {
     private _findCookiesByApi (urls: Url[], key?: string): (Cookie | Cookie[])[] {
         return urls.map(({ domain, path }) => {
             const cookies = key
-                ? this._findCookieSync(domain, path, key)
-                : this._findCookiesSync(domain, path);
+                            ? this._findCookieSync(domain, path, key)
+                            : this._findCookiesSync(domain, path);
 
             return cookies || [];
         });
@@ -285,8 +286,8 @@ export default class Cookies {
         const cookieJarOpts =
                   hostname === LOCALHOST_DOMAIN ||
                   hostname === LOCALHOST_IP ?
-                      { http: true, secure: true } :
-                      { http: true };
+                  { http: true, secure: true } :
+                  { http: true };
         return this._cookieJar.getCookieStringSync(url, cookieJarOpts) || null;
     }
 }

--- a/src/session/cookies.ts
+++ b/src/session/cookies.ts
@@ -62,9 +62,11 @@ export default class Cookies {
         };
     }
 
+    static _isLocalHostDomain = (domain): boolean => domain === LOCALHOST_DOMAIN || domain === LOCALHOST_IP;
+
     static _hasLocalhostDomain (cookie): boolean {
         if (cookie)
-            return cookie.domain === LOCALHOST_DOMAIN || cookie.domain === LOCALHOST_IP;
+            return this._isLocalHostDomain(cookie.domain);
 
         return false;
     }
@@ -111,8 +113,8 @@ export default class Cookies {
 
     setJar (serializedJar): void {
         this._cookieJar = serializedJar
-                          ? CookieJar.deserializeSync(parseJSON(serializedJar))
-                          : new CookieJar();
+            ? CookieJar.deserializeSync(parseJSON(serializedJar))
+            : new CookieJar();
     }
 
     private _convertToExternalCookies (internalCookies: Cookie[]): ExternalCookies[] {
@@ -145,8 +147,8 @@ export default class Cookies {
     private _findCookiesByApi (urls: Url[], key?: string): (Cookie | Cookie[])[] {
         return urls.map(({ domain, path }) => {
             const cookies = key
-                            ? this._findCookieSync(domain, path, key)
-                            : this._findCookiesSync(domain, path);
+                ? this._findCookieSync(domain, path, key)
+                : this._findCookiesSync(domain, path);
 
             return cookies || [];
         });
@@ -283,11 +285,10 @@ export default class Cookies {
         // NOTE: https://github.com/DevExpress/testcafe-hammerhead/issues/2715
         // Cookies with the secure attribute should be passed to the localhost server(without ssl) or any other server(with ssl).
         // CookieJar only checks the protocol, but not a hostname. That is why we should to add secure attribute in case of localhost.
-        const cookieJarOpts =
-                  hostname === LOCALHOST_DOMAIN ||
-                  hostname === LOCALHOST_IP ?
-                  { http: true, secure: true } :
-                  { http: true };
+        const cookieJarOpts = Cookies._isLocalHostDomain(hostname) ?
+            { http: true, secure: true } :
+            { http: true };
+
         return this._cookieJar.getCookieStringSync(url, cookieJarOpts) || null;
     }
 }

--- a/test/server/cookies-test.js
+++ b/test/server/cookies-test.js
@@ -380,6 +380,49 @@ describe('Cookies', () => {
             expect(expectedCookies).eql(cookies);
         });
     });
+    describe('Attach secure cookies to request', () => {
+        beforeEach(() => {
+            cookieJar.setCookies([
+                { name: 'apiCookie1', value: 'value1', domain: 'domain1.com', path: '/', secure: true },
+                { name: 'apiCookie2', value: 'value2', domain: 'localhost', path: '/', secure: true },
+                { name: 'apiCookie3', value: 'value3', domain: '127.0.0.1', path: '/', secure: true },
+            ]);
+        });
+
+        afterEach(() => {
+            cookieJar.deleteCookies();
+            cookieJar._pendingSyncCookies = [];
+        });
+
+        it('Should get secure cookies string from ssl domain', () => {
+            const expectedCookieString = 'apiCookie1=value1';
+            const destInfo             = { url: 'https://domain1.com/', hostname: 'domain1.com' };
+            const cookieStr            = cookieJar.getHeader(destInfo);
+
+            expect(cookieStr).eql(expectedCookieString);
+        });
+
+        it('Should get secure cookies string from localhost', () => {
+            const expectedCookieString1 = 'apiCookie2=value2';
+            const expectedCookieString2 = 'apiCookie3=value3';
+
+            const destInfo1  = { url: 'http://localhost:3006', hostname: 'localhost' };
+            const destInfo2  = { url: 'http://127.0.0.1:3001', hostname: '127.0.0.1' };
+            const cookieStr1 = cookieJar.getHeader(destInfo1);
+            const cookieStr2 = cookieJar.getHeader(destInfo2);
+
+            expect(cookieStr1).eql(expectedCookieString1);
+            expect(cookieStr2).eql(expectedCookieString2);
+        });
+
+        it('Should return null if domain is not HTTPS', () => {
+            const expectedCookieString = null;
+            const destInfo1  = { url: 'http://domain1.com', hostname: 'domain1.com' };
+            const cookieStr1 = cookieJar.getHeader(destInfo1);
+
+            expect(cookieStr1).eql(expectedCookieString);
+        });
+    });
 
     describe('Set cookies', () => {
         afterEach(() => {

